picky-asn1-der/src/application_tag.rs

```diff
@@ -0,0 +1,129 @@
+            }
+        }
+
+        impl<'de, E: de::Deserialize<'de> + Debug + PartialEq> de::Visitor<'de> for Visitor<E> {
+            type Value = E;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str(&format!("A valid DER-encoded ApplicationTag{}", self.1))
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                #[derive(Debug, serde::Deserialize)]
+                struct ApplicationTagInner<V: Debug> {
+                    value: V,
+                }
+
+                let tag_peeker: TagPeeker = seq
+                    .next_element()
+                    .map_err(|e| A::Error::custom(format!("Cannot deserialize application tag: {:?}", e)))?
+                    .ok_or_else(|| A::Error::missing_field("ApplicationTag"))?;
+                let tag = tag_peeker.next_tag;
+
+                if !tag.is_application() {
+                    return Err(A::Error::custom(format!(
+                        "Expected Application class tag but got: {:?}",
+                        tag.class()
+                    )));
+                }
+
+                if tag.number() != self.1 {
+                    return Err(A::Error::custom(format!(
+                        "Expected Application number tag {} but got: {}",
+                        self.1,
+                        tag.number()
+                    )));
+                }
+
+                let rest: ApplicationTagInner<E> = seq
+                    .next_element()
+                    .map_err(|e| A::Error::custom(format!("Cannot deserialize application tag inner value: {:?}", e)))?
+                    .ok_or_else(|| A::Error::missing_field("ApplicationInnerValue"))?;
+
+                Ok(rest.value)
+            }
+        }
+
+        let inner = deserializer
+            .deserialize_enum("ApplicationTag", &["ApplicationTag"], Visitor::<V>::new(T))
+            .map_err(D::Error::custom)?;
+
+        Ok(Self(inner))
+    }
+}
+
+impl<V: ser::Serialize + Debug + PartialEq, const T: u8> ser::Serialize for ApplicationTag<V, T> {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as ser::Serializer>::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        use serde::ser::Error;
+
+        let mut buff = Vec::new();
+        {
+            let mut s = crate::Serializer::new_to_byte_buf(&mut buff);
+            self.0
+                .serialize(&mut s)
+                .map_err(|e| S::Error::custom(format!("Cannot serialize Application tag inner value: {:?}", e)))?;
+        }
+
+        let mut res = vec![Tag::application_constructed(T).inner()];
+
+        Length::serialize(buff.len(), &mut res)
+            .map_err(|e| S::Error::custom(format!("Cannot serialize Length: {:?}", e)))?;
+        res.extend_from_slice(&buff);
+
+        Asn1RawDer(res).serialize(serializer)
```

Comment:
You can use a single buffer

```suggestion
        let mut buf = vec![Tag::application_constructed(T).inner()];

        {
            let mut s = crate::Serializer::new_to_byte_buf(&mut buf);
            self.0
                .serialize(&mut s)
                .map_err(|e| S::Error::custom(format!("Cannot serialize Application tag inner value: {:?}", e)))?;
        }

        Length::serialize(buf.len() - 1, &mut buf)
            .map_err(|e| S::Error::custom(format!("Cannot serialize Length: {:?}", e)))?;
        
        Asn1RawDer(buf).serialize(serializer)
```